### ISSUE TITLE
Using Fable's watch mode together with Webpack's watch mode

### ISSUE THREAD

**Issue by Pauan** [fable-compiler/Fable#432]
I'm trying to use Fable's watch mode combined with Webpack to automatically and efficiently recompile my project whenever any files change. Currently, I'm doing this:

```
// fableconfig.json
{
  "scripts": {
    "postbuild": "webpack"
  }
}
```

When running `fable --watch` it will call `webpack` every time a file changes. This works correctly, but it's quite inefficient: it has to re-run `webpack` from scratch every time, so it doesn't benefit from Webpack's built-in caching. It takes Webpack ~500ms to rebundle my (tiny) project, even if only a single file changed.

I then tried putting `webpack --watch` in the `postbuild` script, but that's absurdly inefficient, because it causes a new `webpack` process to spawn every time a file is changed, and each process never dies, so it just keeps adding more and more processes!

So I then tried removing the `postbuild` script and instead ran this command: `fable --watch & webpack --watch`

This causes `webpack` to run in parallel with `fable`, and it causes Webpack to use its own watching system and caching, so that it can avoid unnecessary work. Now it takes Webpack ~100ms to rebundle my project when a single file changes. That's a huge difference!

However, there's a huge caveat: the `.js` files have to already exist in order for this to work. If this is a fresh build, then the `.js` files don't exist, and therefore Webpack throws an error because it can't find the `.js` files.

So what I want is a way for `fable` to build the project, and after the initial build is complete, then start `webpack --watch` in parallel.

In order to support this, I think it would be better if `postbuild` ran exactly once, after the initial build is complete.

Then you could have a new `postchange` script which is run every time a file changes. By combining `postbuild` and `postchange` you can achieve the same behavior that Fable uses right now.

(And for symmetry, you could also change `prebuild` and add a new `prechange` script)

However, this is a pretty niche request, so I understand if you consider it low priority.

**Comment by alfonsogarciacaro** [comment]
I think this makes a lot of sense and it's not difficult to add. I actually ran into a similar situation and thought of doing the same. The only thing that stopped me is that this change would break existing configurations, so it's probably better to have something like `postbuild_once`. Would that be OK with you? If you have another suggestion for the name, please tell me.

**Comment by Pauan** [comment]
I was hoping that Fable was new enough that breaking changes would be okay, but I completely understand the desire for backwards compatibility.

I'm fine with a new `postbuild_once` option.

**Comment by Pauan** [comment]
I just tested out `fable-compiler` version `0.6.9`, but I ran into a new issue. I'm using this config:

```
{
    "watch": true,
    "scripts": {
        "postbuild-once": "rollup --config --watch"
    }
}
```

Here's what happens:

1. Fable does a full compile of the project.

2. Rollup does a full compile of the output of Fable.

3. I make a change to a file.

4. Fable does not detect the change, instead it does nothing.

It works fine if I remove the `--watch` from Rollup, but then it's slow because it's not using Rollup's watch mode.

It has the same behavior when using Webpack's watch mode.

Perhaps the problem is that Fable is waiting for the `postbuild-once` script to terminate before continuing, but when using Rollup's `--watch` mode, it never terminates.

**Comment by alfonsogarciacaro** [comment]
Ah, you must be right! I'll modify that, thanks for pointing it out!

**Comment by alfonsogarciacaro** [comment]
Can you try with v0.6.10? Also, I'm very interested in Babily and Rollup so it'd really appreciate if you could write a post with your experience developing with Fable and these JS developments tools. BTW, fable-core is also distributed with ES2015 modules, so if you pass the option `--coreLib fable-core/es2015` to Fable CLI, Rollup should be able to strip off the parts of the core lib you don't actually use :+1:

Have fun!

**Comment by Pauan** [comment]
I just tried version `0.6.10`, and it works correctly now!

When using `postbuild` with `rollup --config` it takes ~3,000 milliseconds to recompile a single file.

When using `postbuild-once` with `rollup --config --watch` it takes ~31 milliseconds to recompile a single file.

Rollup's watch mode even manages to detect changes to the FFI files, which is something that Fable doesn't currently do.

Thanks for the advice about `coreLib`: the end result is incredibly small and clean code.

----

I've been developing a `fable-getting-started` repository, which contains all of the code and configuration that a beginner needs to get started with Fable and Rollup. After it's ready, I'll submit a pull request for the samples directory.

**Comment by Pauan** [comment]
Oops, I just realized there's one more issue...

When using `"postbuild-once": "rollup --config --watch"` and Fable's watch mode, if I use `Ctrl+C` to terminate the watching, then everything works fine.

But if I press `Enter` to terminate the watching, it doesn't kill the `rollup` process.

**Comment by tpetricek** [comment]
👍  Sounds like I should remove my hack with sleep that was trying to workaround exactly this issue :) https://github.com/the-gamma/thegamma-script/blob/master/package.json#L12

**Comment by alfonsogarciacaro** [comment]
Wow, a x10 speed up really deserved a new `postbuild-once` option 😄 

Can you try with `fable-compiler@0.6.11`? It should work now. It was a bit tricky to kill the process in Windows: there, Fable prefix the command with `cmd /C` but when I kill the terminal, the subprocess (Webpack, Rollup...) remained alive (there're some solutions to that on the internet but they didn't work on my tests). So I've to run `postbuild-once` scripts without a terminal which unfortunately prevents the execution of `.cmd` or `.bat` files. I hope that's fine.

**Comment by alfonsogarciacaro** [comment]
Also, I'm looking forward to your PR about getting started with Fable and Rollup! :clap:

**Comment by Pauan** [comment]
Actually, it's a 100x speedup, which is even better!

I just tested `0.6.11`, it works correctly now. I'm on Linux so the issue with Windows doesn't affect me.

Thanks a lot!

**Comment by alfonsogarciacaro** [comment]
Great! :tada: Yeah, you're right, I missed a `0` 😉  Thanks to you!

**Comment by rookboom** [comment]
I am having a hard time getting this to fable --watch and webpack --watch to work together on Windows:

Here is my target config from fableconfig.json:
    "targets": {
        "watch": {
            "watch": true,
            "scripts": {
                "postbuild-once": "webpack --watch"
            }
        }

This causes an error (see full trace below).
If I change "postbuild-once" to "postbuild" it works but fable does not actually watch and nothing happens when I change a .fs file.

If I change "webpack --watch" to just "postbuild": "webpack" it works but is slow as mentioned above.

Here is the error I get when using "postbuild-once": 

> todo-elm-fable@1.0.0 watch d:\repos\HelloArchFableSuave
> fable -w --target watch
fable-compiler 0.7.26: Start compilation...
d:\repos\HelloArchFableSuave> npm install
[ERROR] npm
[ERROR]  WARN optional SKIPPING OPTIONAL DEPENDENCY: fsevents@^1.0.0 (node_modules\chokidar\node_modules\fsevents):
npm WARN notsup SKIPPING OPTIONAL DEPENDENCY: Unsupported platform for fsevents@1.0.17: wanted {"os":"darwin","arch":"any"} (current: {"os":"win32","arch":"x64"})
npm WARN todo-elm-fable@1.0.0 No description
npm WARN todo-elm-fable@1.0.0 No repository field.
[WARNING] This expression is a function value, i.e. is missing arguments. Its type is AppMessage<Actions,string> -> unit. (L48,0-L50,8) (d:\repos\HelloArchFableSuave\src\client\client.fs)
Compiled fable-arch\Fable.Arch.Html.js at 2:22:38 PM
Compiled fable-arch\Fable.Arch.App.js at 2:22:38 PM
Compiled fable-arch\Fable.Arch.Virtualdom.js at 2:22:38 PM
Compiled client\client.js at 2:22:38 PM
d:\repos\HelloArchFableSuave> webpack --watch
Watching d:\repos\HelloArchFableSuave\src\client
Press Enter to terminate process.
events.js:160
      throw er; // Unhandled 'error' event
      ^
Error: spawn webpack ENOENT
    at exports._errnoException (util.js:1022:11)
    at Process.ChildProcess._handle.onexit (internal/child_process.js:193:32)
    at onErrorNT (internal/child_process.js:359:16)
    at _combinedTickCallback (internal/process/next_tick.js:74:11)
    at process._tickCallback (internal/process/next_tick.js:98:9)
npm ERR! Windows_NT 10.0.14393
npm ERR! argv "C:\\Program Files\\nodejs\\node.exe" "C:\\Program Files\\nodejs\\node_modules\\npm\\bin\\npm-cli.js" "run" "watch"
npm ERR! node v6.9.4
npm ERR! npm  v3.10.10
npm ERR! code ELIFECYCLE
npm ERR! todo-elm-fable@1.0.0 watch: `fable -w --target watch`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the todo-elm-fable@1.0.0 watch script 'fable -w --target watch'.
npm ERR! Make sure you have the latest version of node.js and npm installed.
npm ERR! If you do, this is most likely a problem with the todo-elm-fable package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     fable -w --target watch
npm ERR! You can get information on how to open an issue for this project with:
npm ERR!     npm bugs todo-elm-fable
npm ERR! Or if that isn't available, you can get their info via:
npm ERR!     npm owner ls todo-elm-fable
npm ERR! There is likely additional logging output above.
npm ERR! Please include the following file with any support request:
npm ERR!     d:\repos\HelloArchFableSuave\npm-debug.log

Watching build tasks has finished.

**Comment by alfonsogarciacaro** [comment]
Hi @rookboom! Are you on Windows? There's a problem with `postbuild-once` on Windows because in order to execute it as a parallel process that can be easily terminated a hidden shell cannot be opened so only `.exe` files can be run (not `.cmd` files as the ones npm creates to run the binary commands in Windows). It's usually fixed by calling `node` directly:

```json
"scripts": {
    "postbuild-once": "node node_modules/webpack/bin/webpack --watch"
}
```

> Please correct the path to `node_modules` folder from `fableconfig.json` if necessary.
